### ISSUE TITLE
Can't prevent duplicates when updating an object in an array

### ISSUE THREAD

**Issue by jambudipa** [Automattic/mongoose#13118]
### Prerequisites

- [X] I have written a descriptive issue title

### Mongoose version

6.9.0

### Node.js version

18.13.0

### MongoDB version

6.0

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

13.2.1

### Issue

I have spent at least six hours of today trying to solve a seemingly easy problem. I am trying to either add or update an object that is in an array of subdocuments, based on the value of a field `price_level`. Here is my schema:
```
import mongoose, { Schema } from 'mongoose';

const orderBookSchema = new Schema({
   pair_name: {
      type: String,
      required: true
   },
   side: String,
   entries: [{ price_level: String, volume: String }],
});

export const OrderBook = mongoose.model('OrderBook', orderBookSchema);
```
If `entries` contains an object that has a particular `price_level`, it should update the `volume` to the new value. Otherwise, add the new element `{ pair_name, volume }` to the array. Here is the code:
```
static updateOrderBook = async (pairName: string, side: 'ask' | 'bid', orderBookEntries: string[][], deleteExisting: boolean = false) => {
   const existingOrderBook = await OrderBook.findOne({ pair_name: pairName, side });

   if (!existingOrderBook) {
      await OrderBook.create({ pair_name: pairName, side, entries: [] });
   }

   if (deleteExisting) {
      const filter = { pair_name: pairName, side };
      const update = { $set: { entries: [] } };
      await OrderBook.updateOne(filter, update).exec();
   }

   for await (const [priceLevel, volume] of orderBookEntries) {
      const filter = {
         pair_name: pairName,
         side,
         'entries.price_level': priceLevel
      };

      const update = { $set: { 'entries.$.volume': volume } };
      const result = await OrderBook.updateOne(filter, update).exec();

      if (result.matchedCount === 0) {
         const newElement = { price_level: priceLevel, volume: volume };
         const addUpdate = { $addToSet: { entries: newElement } };
         await OrderBook.updateOne({ pair_name: pairName, side }, addUpdate).exec();
      }
   }

   const filter = { pair_name: pairName, 'entries.volume': '0.00000000' };
   const update = { $pull: { entries: { volume: '0.00000000' } } };
   await OrderBook.updateOne(filter, update).exec();
};
```
I have tried so many different permutations of operators and so on, and checked the properties of `result`, but whatever I do, a duplicate or two slips through. It is not consistent – mostly it seems to behave, and when I hardcode some values to test, it is fine.

But occasionally, when it's coming from the Kraken web socket API, duplicates appear.

ChatGPT is doing my head in. It has suggested so many different things, and of those that compiled, they did not work.

I have tried:
- using a `Number` instead of a `String` for the subdocument properties
- applying a transaction to the code block
- removed replication
- not using `.exec()` at the end of the update function calls

It's really tricky to debug, because it is intermittent and I am unable to reproduce it by manually passing in mock data.

Is there a fault with my query? This sort of thing should not happen:
```
entries: [
    // ...more
    {
      price_level: '23463.40000',
      volume: '0.13447916',
      _id: new ObjectId("6401151cc708de7a1ff9bf01")
    },
    {
      price_level: '23463.40000',
      volume: '0.13447916',
      _id: new ObjectId("6401151cc708de7a1ff9c0ac")
    }
    // ...more
]
```
...but it does.

**Comment by jambudipa** [comment]
So I rejigged the schema, and it turned out to be agonisingly simple:
```
static updateOrderBook = async (
   pairName: string,
   side: 'ask' | 'bid',
   orderBookEntries: string[][],
   deleteExisting: boolean = false
) => {
   const orderBook = { pairName, side };

   if (deleteExisting) {
      await OrderBook.deleteMany(orderBook);
   }

   for await (const [priceLevel, volume] of orderBookEntries) {
      const filter = { ...orderBook, priceLevel };
      if (+volume === 0) {
         await OrderBook.findOneAndDelete(filter);
      } else {
         await OrderBook.findOneAndUpdate(
            filter,
            { $set: { volume } },
            { upsert: true }
         );
      }
   }
};
```

**Comment by vkarpov15** [comment]
So instead of having one order book, you have a separate document for every price level in the order book? That's an interesting idea. It definitely works. Just a bit slower to load the current state of the order book.

Another potential approach would be to use a map rather than an array, with price levels as the key. Just make sure to multiply out the price level to avoid dots in the key name.

```
const mongoose = require('mongoose');

const orderBookSchema = new mongoose.Schema({
   pair_name: {
      type: String,
      required: true
   },
   side: String,
   entries: { type: Map, of: Number }
});

const OrderBook = mongoose.model('OrderBook', orderBookSchema);

// Update a given price level
OrderBook.updateOne({ _id: orderBookId }, { $inc: { `entries.${priceLevel}`: volume } });
```

**Comment by jambudipa** [comment]
It’s ok, I rejigged my schema and now it is working perfectly. As you mention, speed might have been a concern but I have tested it and it’s easily fast enough. There are only a couple of thousand documents or so.Perhaps I can try out your Map suggestion if speed becomes an issue.Sent from my iPhoneOn 4 Mar 2023, at 18:54, Valeri Karpov ***@***.***> wrote:﻿
So instead of having one order book, you have a separate document for every price level in the order book? That's an interesting idea. It definitely works. Just a bit slower to load the current state of the order book.
Another potential approach would be to use a map rather than an array, with price levels as the key. Just make sure to multiply out the price level to avoid dots in the key name.
const mongoose = require('mongoose');

const orderBookSchema = new mongoose.Schema({
   pair_name: {
      type: String,
      required: true
   },
   side: String,
   entries: { type: Map, of: Number }
});

const OrderBook = mongoose.model('OrderBook', orderBookSchema);

// Update a given price level
OrderBook.updateOne({ _id: orderBookId }, { $inc: { `entries.${priceLevel}`: volume } });
